docs/discussion/type-safety.md

```diff
@@ -0,0 +1,74 @@
+---
+title: Type safety
+---
+
+# Type safety
+
+React Router generates types for each route in your app that you can use to get type safety for each route module export.
+
+For example, let's say you have a `products/:id` route configured:
+
+```ts filename=app/routes.ts
+import {
+  type RouteConfig,
+  route,
+} from "@react-router/dev/routes";
+
+export const routes: RouteConfig = [
+  route("products/:id", "./routes/product.tsx"),
+];
+```
+
+Then, you can import route-specific types like so:
+
+```tsx filename=app/routes/product.tsx
+import type * as Route from "./+types.product";
+// types generated for this route 👆
+
+export function loader({ params }: Route.LoaderArgs) {
+  //                      👆 { id: string }
+  return { planet: `world #${params.id}` };
+}
+
+export default function Component({
+  loaderData, // 👈 { planet: string }
+}: Route.ComponentProps) {
+  return <h1>Hello, {loaderData.planet}!</h1>;
+}
+```
+
+If you haven't done so already, check out our guide for [setting up type safety][setting-up-type-safety] in a new project.
+
+## `typegen` command
+
+You can manually generate types with the `typegen` command:
+
+```sh
+react-router typegen
+```
+
+You can also use `--watch` to automatically regenerate types as files change:
+
+```sh
+react-router typegen --watch
+```
+
+The following types are generated for each route:
+
+- `LoaderArgs`
+- `ClientLoaderArgs`
+- `ActionArgs`
+- `ClientActionArgs`
+- `HydrateFallbackProps`
+- `ComponentProps` (for the `default` export)
+- `ErrorBoundaryProps`
```

Comment:
This feels more reference-y than guide-y, but CLI doesn't have automatically generated reference docs so felt ok to include here